[doc] add benchmark.md (#122)
megaease/easegress: doc/benchmark.md

	include /etc/nginx/sites-enabled/*;
}

server {
    listen       8080;
    server_name  localhost;

    location /pipeline {
        proxy_pass     http://${echo_svr_ip}:9095;
        keepalive_timeout  60;
    }
}
```

**Easegress**
1. Easegress config
``` yaml
name: member-001
cluster-name: cluster-test
cluster-role: writer
cluster-client-url: http://127.0.0.1:2379
cluster-peer-url: http://127.0.0.1:2380
cluster-join-urls:
api-addr: 127.0.0.1:2381
data-dir: ./data
wal-dir: ""
cpu-profile-file:
memory-profile-file:
log-dir: ./log
member-dir: ./member
std-log-level: INFO
```

2.HTTPServer+Pipeline

``` yaml
- filters:
  - kind: Proxy
    mainPool:
      loadBalance:
        policy: roundRobin
      servers:
      - url: http://${echo_server_ip}:9095
    name: proxy
  flow:
  - filter: proxy
    jumpIf: {}
  kind: HTTPPipeline
  name: pipeline-demo

- http3: false
  https: false
  keepAlive: true
  keepAliveTimeout: 60s
  kind: HTTPServer
  maxConnections: 10240
  name: server-demo
  port: 10080
  rules:
  - host: ""
    hostRegexp: ""
    paths:
    - backend: pipeline-demo
      headers: []
      pathPrefix: /pipeline
      rewriteTarget: ""
```

3. HTTPPipeline for baseline

> * In order to comparing with NGINX's index.html API. Pipeline uses ResponseAdaptor to reply a NGINX'index.html like response.(Yeah, replace `nginx` with `easegress` already)

Reviewer: ```suggestion
* In order to compare with NGINX's index.html API. The pipeline uses ResponseAdaptor to reply an NGINX' index.html like response.(Yeah, replace `nginx` with `easegress` already)
```

          }
      </style>
      </head>
      <body>
      <h1>Welcome to easegress!</h1>
      <p>If you see this page, the easegress web server is successfully installed and working. Further configuration is required.</p>

      <p>For online documentation and support please refer to <a href="http://easegress.org/">easegress.org</a>.<br/> Commercial support is available at <a href="http://easegress.com/">easegress.com</a>.</p>
      <p><em>Thank you for using easegress.</em></p>
      </body>
      </html>
```

**Traefik**
1. Running binary directly with command `./traefik -c ./traefik.yml` [1]
2. Static lonfig: traefik.yaml
``` yaml 
log:
  level: INFO

entryPoints:
  web:
    address: ":8081"

providers:
  file:
    filename: /${file_path}/dynamic_conf.yml

```
3. Dynamic config: dynamic_conf.yml
``` yaml
http:
  routers:
    my-router:
      rule: "PathPrefix(`/pipeline`)"
      service: foo
      entryPoints:
      - web

  services:
    foo:
      loadBalancer:
        servers:
        - url: "http://${echo_server_ip}:9095"
```

**Echo Server**
1. [Source code](https://github.com/megaease/easegress/tree/main/example/backend-service/mirror)

> 2. Its logic is accepting HTTP request and printing to the console, in this testing, we only uses `9095` port.

Reviewer: ```suggestion
2. Its logic is accepting HTTP request and printing to the console, in this testing, we only use `9095` port.
```

## Testing 
### Baseline test
1. **Echo-server**
* Loading echo-server directly from `vm02` to `vm03`
* **Scenario 1**: 50 concurrency/900 requests/2 miniutes limitation/not QPS limitation

``` bash

./hey -n 900   -c 50  -m GET http://${vm03_ip}:9095/pipeline -z 2m    

```

* **Scenario 2**: 100 concurrency/90000 requests/2 miniutes limitation/not QPS limitation
``` bash

./hey -n 90000  -c 100  -m GET http://${vm03_ip}:9095/pipeline -z 2m   

```

* **Scenario 3**: 120 concurrency/90000 requests/ 2 miniutes limitation/not QPS limitation

``` bash

> ./hey -n 90000  -c 120  -m GET http://${vm03_ip}:9095/pipeline -z 2m   

Reviewer: in `hey` command line, when `-z` is specified, `-n` is ignored.

Author: Thanks for reminding, but in our test env

``` bash
./hey -n 90000  -c 120 -m GET http://${ip}:8080/index.html -z 2m    
Summary:
 Total:	5.8363 secs
 Slowest:	0.0913 secs
 Fastest:	0.0001 secs
 Average:	0.0075 secs
 Requests/sec:	15420.7851
Response time histogram:
 0.000 [1]	|
 0.009 [59076]	|■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■
 0.018 [22476]	|■■■■■■■■■■■■■■■
 0.027 [6389]	|■■■■
 0.037 [1391]	|■
 0.046 [483]	|
 0.055 [94]	|
 0.064 [43]	|
 0.073 [34]	|
 0.082 [2]	|
 0.091 [11]	|
Latency distribution:
 10% in 0.0004 secs
 25% in 0.0012 secs
 50% in 0.0053 secs
 75% in 0.0114 secs
 90% in 0.0179 secs
 95% in 0.0222 secs
 99% in 0.0343 secs
Details (average, fastest, slowest):
 DNS+dialup:	0.0000 secs, 0.0000 secs, 0.0402 secs
 DNS-lookup:	0.0000 secs, 0.0000 secs, 0.0000 secs
 req write:	0.0001 secs, 0.0000 secs, 0.0436 secs
 resp wait:	0.0036 secs, 0.0001 secs, 0.0618 secs
 resp read:	0.0037 secs, 0.0000 secs, 0.0602 secs
Status code distribution:
 [200]	90000 responses
```
It won't affect the total request number and the testing result,  if we put `-z ` at the beginning, it will ignore the `-n` setting. 
But it does confuse readers. So let me update these commands.

```

| Scenario | Total | Slowest | Fastest | Average | RPS   | 90% Latency | 95% Latency | 99% Latency | load average(top -c ) |
| -------- | ----- | ------- | ------- | ------- | ----- | ----------- | ----------- | ----------- | --------------------- |
| #1       | 0.19s | 0.015s  | 0.0103s | 0.0109s | 4517  | 0.0119s     | 0.0126s     | 0.0144s     | 0/0/0                 |
| #2       | 9.88s | 0.054s  | 0.0101s | 0.0109s | 9109  | 0.0118s     | 0.0124s     | 0.0138s     | 0.96/0.42/0.19        |
| #3       | 8.48s | 0.042s  | 0.0101s | 0.0110s | 10768 | 0.0122s     | 0.0129s     | 0.0149s     | 1.34/0.46/0.20        |

2. **Nginx**

> * Loading Nginx's `index.html` url, from `vm02` to `vm01`.

Reviewer: There are `Nginx`, `NGINX`, and `nginx` in the doc, propose to use the same style.

Author: OK, I will update `Nginx` to `NGINX`. Since some `nginx` format usages are unavoidable, e.g., `pid /run/nginx.pid;`, I will keep them as `nginx`.
